psa-crypto-sys/build.rs

```diff
@@ -141,9 +141,11 @@ fn main() -> Result<()> {
         statically = true;
     }
 
-    compile_shim_library(include.clone())?;
-    generate_mbed_crypto_bindings(include)?;
-    link_to_lib(lib, statically)?;
+    generate_mbed_crypto_bindings(include.clone())?;
```

Comment:
Tried that, didn't work - the constants depend on the shim bindings being generated, all of them are equal to `shim_<something something>` - plus, the header files are supposed to be implementation independent - we're not compiling or linking a library, simply pulling in the C-defined interface and generating bindings to it.

Reply:
> Tried that, didn't work - the constants depend on the shim bindings being generated, all of them are equal to shim_<something something>

Ideally, we would define the constants which values are indicated in the specification ourselves, the same way as we define the types ourselves. That way we do not have to rely on the `vendor` folder _at all_ when generating an implementation-independant build. I did not check the context of the change in that file, but basically I think this whole build script should not be executed when the `implementation-defined` feature is not on: everything is self dependant. That way, we can also remove from the shimmed constants the ones that are specified.

> the header files are supposed to be implementation independent - we're not compiling or linking a library, simply pulling in the C-defined interface and generating bindings to it.

Agree for the parts of the header that are defining the specified types and constants, but not everything. We could also chose to generate bindings all the time and only re-surface from them the types and constants (fron the shim) that are specified, but that seems less clean than re-defining them ourselves and not needing bindgen at all.

Reply:
I think we've drifted off from how a `-sys` crate should work and made it non-intuitive. The purpose of this kind of crate is generally to expose the interface and link to the crate. If we just define everything here we're not really doing any of those two. Hence, I'd probably lean towards generating all the bindings and exporting the types we think are important, but having a `link` feature (that is enabled by default) and which, when turned off, leaves linking to the user. The rest of the build would work the same way.

Reply:
It might be cumbersome to do it by hand but we can just copy the ones from one bindgen generation 😃

Reply:
> I think we've drifted off from how a -sys crate should work and made it non-intuitive. The purpose of this kind of crate is generally to expose the interface and link to the original library.

I disagree, we do exactly the same as before but instead of letting `bindgen` generate the constants and types, we define them ourselves. We can do that because they are specified and stable in the Crypto API. The constants and types we define can still be used for FFI interactions with a PSA Crypto implementation and so we are filling that objective.

Maybe the `link` name is better than `implementation-defined`. When it is not set, this crate only generates what it can from the specified types and constants of the API.